lib/Runtime/Library/JSON.cpp

```diff
@@ -688,55 +689,45 @@ namespace JSON
 
                         result = Js::ConcatStringBuilder::New(this->scriptContext, propertyCount);
                         Js::DynamicTypeHandler * typeHandler = dynamicObject->GetTypeHandler();
-                        enumerator.Reset();
-                        Js::PropertyId previousId;
+
                         // if object has an objectarray, (aka indexed properties)
                         // we need to loop them first since they won't show up on the second loop below
                         if (dynamicObject->HasObjectArray())
                         {
-                            int totalNumPropertyCount = this->GetPropertyCount(object, &enumerator) - propertyCount;
-                            Assert(totalNumPropertyCount >= 0);
-
-                            int index = 0;
                             enumerator.Reset();
-                            while (index < totalNumPropertyCount && (propertyName = enumerator.MoveAndGetNext(id)) != NULL)
+                            while ((propertyName = enumerator.MoveAndGetNext(id)) != NULL)
                             {
                                 // if unsuccessful get propertyId from the string
-                                scriptContext->GetOrAddPropertyRecord(propertyName->GetString(), propertyName->GetLength(), &propRecord);
+                                scriptContext->GetOrAddPropertyRecord(propertyName, &propRecord);
+                                if (!propRecord->IsNumeric()) break;
                                 id = propRecord->GetPropertyId();
                                 StringifyMemberObject(propertyName, id, value, (Js::ConcatStringBuilder*)result, indentString, memberSeparator, isFirstMember, isEmpty);
-                                index++;
                             }
                         }
 
                         // loop over the named properties
                         for (uint32 i = 0; i < propertyCount; i++)
                         {
                             id = typeHandler->GetPropertyId(scriptContext, (Js::PropertyId)i);
```

Comment:
This line has confused me for a while, and looking into it more, I realized why -- the second argument, if its going to be casted, should be casted to a `Js::PropertyIndex` rather than a `Js::PropertyId`